linkerd/stack/src/map_err.rs

```diff
@@ -1,58 +1,263 @@
-        super::layer::mk(move |inner| Self::new(inner, f.clone()))
+impl<S> MapErr<(), S> {
+    pub fn layer_boxed() -> impl layer::Layer<S, Service = Self> + Clone {
+        MapErr::layer(())
     }
 }
 
-impl<Req, E, S, F: Clone> super::Service<Req> for MapErr<S, F>
+impl<Req, W, S> super::Service<Req> for MapErr<W, S>
 where
     S: super::Service<Req>,
-    F: FnOnce(S::Error) -> E,
+    W: WrapErr<S::Error> + Clone,
 {
     type Response = S::Response;
-    type Error = E;
-    type Future = futures::future::MapErr<S::Future, F>;
+    type Error = W::Error;
+    type Future = ResponseFuture<W, S::Future>;
 
     #[inline]
-    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), E>> {
-        self.inner.poll_ready(cx).map_err(self.f.clone())
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(|e| self.wrap.wrap_err(e))
     }
 
     #[inline]
     fn call(&mut self, req: Req) -> Self::Future {
-        self.inner.call(req).map_err(self.f.clone())
+        let inner = self.inner.call(req);
+        ResponseFuture {
+            inner,
+            wrap: self.wrap.clone(),
+        }
     }
 }
 
-impl<Req, E, P, S, F: Clone> super::Proxy<Req, S> for MapErr<P, F>
+impl<Req, W, P, S> super::Proxy<Req, S> for MapErr<W, P>
 where
+    W: WrapErr<P::Error> + Clone,
     P: super::Proxy<Req, S>,
     S: super::Service<P::Request>,
-    F: FnOnce(P::Error) -> E,
-    E: Into<Error>,
 {
     type Request = P::Request;
     type Response = P::Response;
-    type Error = E;
-    type Future = futures::future::MapErr<P::Future, F>;
+    type Error = W::Error;
+    type Future = ResponseFuture<W, P::Future>;
 
     #[inline]
     fn proxy(&self, inner: &mut S, req: Req) -> Self::Future {
-        self.inner.proxy(inner, req).map_err(self.f.clone())
+        let inner = self.inner.proxy(inner, req);
+        ResponseFuture {
+            inner,
+            wrap: self.wrap.clone(),
+        }
+    }
+}
+
+// === impl ResponseFuture ===
+
+impl<T, E, W, F> Future for ResponseFuture<W, F>
+where
+    W: WrapErr<E>,
+    F: Future<Output = Result<T, E>>,
+{
+    type Output = Result<T, W::Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        this.inner
+            .poll(cx)
+            .map_err(|error| this.wrap.wrap_err(error))
+    }
+}
+
+// === impl WrapFromTarget ===
+
+impl<T, InE, OutE> WrapErr<InE> for WrapFromTarget<T, OutE>
+where
+    InE: Into<Error>,
+    OutE: for<'a> From<(&'a T, Error)>,
+    OutE: Into<Error>,
+{
+    type Error = OutE;
+
+    fn wrap_err(&self, error: InE) -> OutE {
+        OutE::from((&self.target, error.into()))
+    }
+}
+
+impl<T: Clone, E> Clone for WrapFromTarget<T, E> {
+    fn clone(&self) -> Self {
+        Self {
+            target: self.target.clone(),
+            _err: PhantomData,
+        }
```

Comment:
nit: should probably have a `Debug` impl as well, since otherwise it's going to break the `NewService`'s `Debug` impl?